# Introduction:
https://app.dataquest.io/m/356/guided-project%3A-exploring-hacker-news-posts/1/introduction


## Rows all have:
* id 
* title 
* url 
* num_points 
* num_comments 
* author 
* created at

## Looking at titles that begin with "Ask HN" to ask questions and "Show HN" to show projects or interesting things to the community


# Answer the following questions:
1. Do Ask HN or Show HN receive more comments on average
2. Do posts created at a certain time receive more comments on average?

In [2]:
opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(opened_file)
posts_data = list(read_file)

print(posts_data[0])#Prints the header row
len(posts_data) #Prints the length to make sure all rows are displayed


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


20101

In [3]:
hn = posts_data
print(hn[0:5]) #Prints the first five rows of hn

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [4]:
header = hn[0] #Sets header to another variable
print(hn[0:5]) #Shows list before deletion of header
del hn[0] #Extracts the header
print(hn[0:5]) #Displays list after deletion of header

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck U

# Filtering data: We want to specifically analyze ask hn and show hn post so we will filter for those categories

In [5]:
#Create three empty lists
ask_posts = []
show_posts = []
other_posts = []

#Loop through and sort into the three list
for row in hn:
    title = row[1]
    title = title.lower() #Makes ;all titles lowercase to search for a certain string because capitalization matters
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))   
print(len(show_posts))        
print(len(other_posts))        

1744
1162
17194


## We have:

* 1744 Ask Posts
* 1162 Show Posts

In [6]:
print(ask_posts[0:5])#First 5 ask posts

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [7]:
print(show_posts[0:5])#First 5 show posts

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


## Now we will determine if show or ask post have more comments on average

In [8]:
total_ask_comments = 0
for row in ask_posts:
    comment = row[4]
    comment = int(comment) #Convert string to int 
    total_ask_comments += comment #Adds the value of each comment to the total
    
print(total_ask_comments) #Prints total amount of Ask comments

#Find average number of comments on asks comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

24483
14.038417431192661


# We have 24483 Ask Posts total at an average of 14 comments per Ask Post

In [9]:
total_show_comments = 0
for row in show_posts:
    comment = row[4]
    comment = int(comment)#Convert string to int
    total_show_comments += comment #Adds the value of each comment to the total

print(total_show_comments) #Prints total amount of Show comments

#Find average number of comments on show comments
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

11988
10.31669535283993


# We have 11988 Show Posts total at an average for 10 comments per Show Post

# On average Ask Post receive more comments with 14 comments per post compared to Show Posts with 10 comments per post

## We will pivot our analysis to Asks Post.
## Our next step is to determine if ask posts created at a certain time are more likely to attract comments

In [10]:
# Calculate the amount of asks posts created in each hour of the da
#along with the number of comments received

import datetime as dt
result_list = [] #List of lists
for row in ask_posts:
    created_at = row[6]
    number_of_comments = row[4]
    number_of_comments = int(number_of_comments)
    result_list.append([created_at, number_of_comments]) #Creates list of lists
print(result_list[0]) #Displays the list



['8/16/2016 9:55', 6]


In [11]:
#Calculate the average number of comments ask post 
#receive by hour created
counts_by_hour = {}

comments_by_hour = {}

for row in result_list:
    hour = row[0]
    hour = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M")
    hour = hour.hour #Extract the %H from above and put it in the Hour variable
    comments_at_hour = row[1]
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1 #Counts the number of posts in at an hour of day
        comments_by_hour[hour] = comments_at_hour  #Counts the number of comments at each hour
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments_at_hour


## Contains the number of ask posts created during each hour of day

In [12]:
print(counts_by_hour) #Prints how many posts were posted at each hour


{0: 55, 1: 60, 2: 58, 3: 54, 4: 47, 5: 46, 6: 44, 7: 34, 8: 48, 9: 45, 10: 59, 11: 58, 12: 73, 13: 85, 14: 107, 15: 116, 16: 108, 17: 100, 18: 109, 19: 110, 20: 80, 21: 109, 22: 71, 23: 68}


## Contains the number of comments under ask posts at each hour received

In [13]:
print(comments_by_hour) #Prints the number of comments for each hour of the day


{0: 447, 1: 683, 2: 1381, 3: 421, 4: 337, 5: 464, 6: 397, 7: 267, 8: 492, 9: 251, 10: 793, 11: 641, 12: 687, 13: 1253, 14: 1416, 15: 4477, 16: 1814, 17: 1146, 18: 1439, 19: 1188, 20: 1722, 21: 1745, 22: 479, 23: 543}


## 1. We need to calculate the average number of comments for posts created during each hour of the day

In [14]:
#We need to check the average comments for post for each hour
#Ex how many comments for posts at midnight 

avg_by_hour = [[h, comments_by_hour[h]/counts_by_hour[h]] for h in counts_by_hour]

print(avg_by_hour)


[[0, 8.127272727272727], [1, 11.383333333333333], [2, 23.810344827586206], [3, 7.796296296296297], [4, 7.170212765957447], [5, 10.08695652173913], [6, 9.022727272727273], [7, 7.852941176470588], [8, 10.25], [9, 5.5777777777777775], [10, 13.440677966101696], [11, 11.051724137931034], [12, 9.41095890410959], [13, 14.741176470588234], [14, 13.233644859813085], [15, 38.5948275862069], [16, 16.796296296296298], [17, 11.46], [18, 13.20183486238532], [19, 10.8], [20, 21.525], [21, 16.009174311926607], [22, 6.746478873239437], [23, 7.985294117647059]]


## We need to change our data so that it is easy to read

In [15]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]]) #Swap the columns so that average is first then hour
print(swap_avg_by_hour)

[[8.127272727272727, 0], [11.383333333333333, 1], [23.810344827586206, 2], [7.796296296296297, 3], [7.170212765957447, 4], [10.08695652173913, 5], [9.022727272727273, 6], [7.852941176470588, 7], [10.25, 8], [5.5777777777777775, 9], [13.440677966101696, 10], [11.051724137931034, 11], [9.41095890410959, 12], [14.741176470588234, 13], [13.233644859813085, 14], [38.5948275862069, 15], [16.796296296296298, 16], [11.46, 17], [13.20183486238532, 18], [10.8, 19], [21.525, 20], [16.009174311926607, 21], [6.746478873239437, 22], [7.985294117647059, 23]]


In [16]:
swap_avg_by_hour = sorted(avg_by_hour, key=lambda row:row[1], reverse=True)
print(swap_avg_by_hour) #Prints highest to lowest


[[15, 38.5948275862069], [2, 23.810344827586206], [20, 21.525], [16, 16.796296296296298], [21, 16.009174311926607], [13, 14.741176470588234], [10, 13.440677966101696], [14, 13.233644859813085], [18, 13.20183486238532], [17, 11.46], [1, 11.383333333333333], [11, 11.051724137931034], [19, 10.8], [8, 10.25], [5, 10.08695652173913], [12, 9.41095890410959], [6, 9.022727272727273], [0, 8.127272727272727], [23, 7.985294117647059], [7, 7.852941176470588], [3, 7.796296296296297], [4, 7.170212765957447], [22, 6.746478873239437], [9, 5.5777777777777775]]


In [42]:
top_three = []
counter = 0 

#Use str.format to print the hour and average
for row in swap_avg_by_hour:
    hour = str(row[0])
    average = row[1]
    date = dt.datetime.strptime(hour,"%H")
    hour = date.hour
    hour_string ="{}:00: {:.2f} average comments per post".format(hour, average)
    print(hour_string)
    if counter <3:
        top_three.append([hour_string])
        counter += 1
    #Add top three times to a new list



15:00: 38.59 average comments per post
2:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00: 11.46 average comments per post
1:00: 11.38 average comments per post
11:00: 11.05 average comments per post
19:00: 10.80 average comments per post
8:00: 10.25 average comments per post
5:00: 10.09 average comments per post
12:00: 9.41 average comments per post
6:00: 9.02 average comments per post
0:00: 8.13 average comments per post
23:00: 7.99 average comments per post
7:00: 7.85 average comments per post
3:00: 7.80 average comments per post
4:00: 7.17 average comments per post
22:00: 6.75 average comments per post
9:00: 5.58 average comments per post


In [70]:
#print top three average comments per post
for row in top_three:
    #print(row)
    time = row[0]
    print(time)

15:00: 38.59 average comments per post
2:00: 23.81 average comments per post
20:00: 21.52 average comments per post


# Posts created at the following times had the most comments on average:
* 3:00 pm
* 2:00 am
* 8:00 pm